### PR TITLE
Utilities

### PR DESCRIPTION
**Pull request by eboyda**
Merge rebuild.py into master.

### CODE REVIEW

File: code/rebuild.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    output_neurons = range(classifier.params[-1].get_value().shape[0])
+    return numpy.argmax(fire_neurons(input,output_neurons,net),axis=1)
+
+
+def logreg_validate(input,labels,classifier):
+    """Compute the fractional zero-one loss for the classifer on input, given
+    labels."""
+    # WIP?  It seems to work but returns slightly different
+    # numbers on MNIST than seen during training: On val set: .0743 vs
+    # .075 during training; on test set: .0775 vs .0749 during training.
```
Comment (robinkraft) [inline code/rebuild.py]: Have you tried 0.0765432?

File: code/rebuild.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+def save_layer(outfile,W,b):
+    """Save weights and biases for a single NN layer."""
+    with open(outfile,'wb') as of:
+        cPickle.dump(W.get_value(borrow=True),of,-1)
+        cPickle.dump(b.get_value(borrow=True),of,-1)
+    return
+
+def save_params(file_prefix,params):
+    """Save weights and biases for all NN layers, one layer per file."""
+    for layer in range(len(params)/2):
```
Comment (robinkraft) [inline code/rebuild.py]: is this `2` a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming))? If so, better to pull it out into its own variable.

File: code/rebuild.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        outfile = file_prefix+'Layer'+str(layer)+'.pkl'
+        save_layer(outfile,*params[2*layer:2*(layer+1)])
```
Comment (robinkraft) [inline code/rebuild.py]: ditto re: magic numbers

File: code/rebuild.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        cPickle.dump(classifier, f)
+    return
+
+### REBUILD ###
+
+def load_params(file):
+    """Load pickled weights and biases for a single NN layer.  Assumes weights
+    and biases were pickled in sequence into the specified file.
+    """
+    W = theano.shared(
```
Comment (robinkraft) [inline code/rebuild.py]: I assume `W` stands for weights? I'm generally not a fan of one-letter variable names but there are legit reasons, including here probably. So I thought I'd flag this and `b`. Here's a quick comment on why this jumped out at me:

http://programmers.stackexchange.com/a/71720

    A short variable name which represents its intent clearly is to be preferred over a big variable name.

So if you've got small scope of usage, or large scope with clear intent, then it's all good. It's kind of a religious thing though - there's no right answer. No need to change.

Comment (matthewhanson) [inline code/rebuild.py]: Even in the case where a single letter accurately describes something (e.g, using same names as in well known equations such as c for speed of light), all it takes is one attempt to grep for it in a tree of files to see it's not a great idea no matter what.

**Comment by eboyda** [comment]
Trial by fire!  I renamed the offending variables, rewrote the function with potential magic number to be more pythonic, and improved formatting per python style guide.  Check it out.  Thanks, guys.

File: code/rebuild.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        )
+    return fire()[:,neurons]
+
+def logreg_predict(inputvecs,classifier):
+    """Return the predicted output labels for input vectors."""
+    output_neurons = range(classifier.params[-1].get_value().shape[0])
+    return numpy.argmax(fire_neurons(inputvecs,output_neurons,net),axis=1)
+
+
+def logreg_validate(input,labels,classifier):
```
Comment (robinkraft) [inline code/rebuild.py]: looks like pure `input` is still around ...

File: code/rebuild.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    """Compute the fractional zero-one loss for the classifer on inputvecs,
+    given labels."""
+    # WIP?  It seems to work but returns slightly different
+    # numbers on MNIST than seen during training: On val set: .0743 vs
+    # .075 during training; on test set: .0775 vs .0749 during training.
+    loss = theano.function(
+        inputs=[],
+        outputs=classifier.errors(labels),
+        givens={classifier.input:input}
```
Comment (robinkraft) [inline code/rebuild.py]: here too

File: code/rebuild.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    # WIP: Though this code is ripped from logistic_sgd.py, I haven't
+    # been able to reload the pickled classifier.  (See
+    # load_classifier(), below.)
+    with open(outfile,'wb') as of:
+        cPickle.dump(classifier, f)
+    return
+
+### REBUILD ###
+
+def load_params(file):
```
Comment (robinkraft) [inline code/rebuild.py]: These `file` variables need to be renamed too.

File: code/rebuild.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    )
+    b = theano.shared(
+        value=numpy.zeros(
+            (0,),
+            dtype=theano.config.floatX
+        ),
+        name='b',
+        borrow=True
+    )
+    with open(file,'rb') as inf:
```
Comment (robinkraft) [inline code/rebuild.py]: here.

**Comment by robinkraft** [comment]
@eboyda I'm still seeing some random `W` and `b` variables around. Is that intentional? I just want to be sure they're not just slipping through the cracks.

**Comment by eboyda** [comment]
Robin, I'm using W and b to comport with theano conventions, though they use them as such only as class attributes (i.e. self.W, self.b).  When they are simple variables would you prefer that I call them weights and biases instead?

**Comment by robinkraft** [comment]
Naw it's all good - if there's a reason for the names then that's cool.

Just be sure to fix up the remaining `file` and `input` variables.

**Comment by eboyda** [comment]
Thanks, and will do.  I think I lost some edits switching git branches... but I'm getting the hang of it.

**Comment by robinkraft** [comment]
Bummer about the edits ... `git stash` may be your new best friend. Let's you temporarily set aside edits and get them back.

http://www.gitguys.com/topics/temporarily-stashing-your-work/

**Comment by eboyda** [comment]
Cool tip with git stash, thx.  I was surprised to learn that I can have files from two different branches open at the same time, but saving them doesn't necessarily work!

**Comment by robinkraft** [comment]
Merging. I'll bug you about tests another time :)

**Pull request merged**